### PR TITLE
Tesseract: add with-training-tools and with-opencl options

### PR DESCRIPTION
**Pull request by ryanfb**

### CODE REVIEW

File: Library/Formula/tesseract.rb [hunk 1]
```diff
@@ -24,10 +25,21 @@ class Tesseract < Formula
   end
 
   option "all-languages", "Install recognition data for all languages"
+  option "with-training-tools", "Install OCR training tools"
+  option "with-opencl", "Enable OpenCL support"
 
   depends_on "libtiff" => :recommended
   depends_on "leptonica"
 
+  if build.with? "training-tools"
```
Comment (DomT4) [inline Library/Formula/tesseract.rb]: How widely used is this likely to be? I'm a little bleh about introducing so many optional dependencies here unless it's going to prove popular/widely useful/etc.

File: Library/Formula/tesseract.rb [hunk 2]
```diff
@@ -17,6 +17,7 @@ class Tesseract < Formula
     depends_on "automake" => :build
     depends_on "libtool" => :build
     depends_on "pkg-config" => :build
+    depends_on "cairo" => :build
```
Comment (DomT4) [inline Library/Formula/tesseract.rb]: Why is this necessary now, but not previously?

File: Library/Formula/tesseract.rb [hunk 3]
```diff
@@ -58,8 +70,16 @@ def install
     ENV.cxx11
 
     system "./autogen.sh" if build.head?
-    system "./configure", "--disable-dependency-tracking", "--prefix=#{prefix}"
+    if build.with? "opencl"
```
Comment (DomT4) [inline Library/Formula/tesseract.rb]: Just use an `args` array here, rather than duplicating the whole block.

```ruby
args = %W[
  --disable-dependency-tracking
  --prefix=#{prefix}
]

args << "--enable-opencl" if build.with? "opencl"

system "./configure", *args
```

File: Library/Formula/tesseract.rb [hunk 4]
```diff
@@ -58,8 +70,16 @@ def install
...
+      system "./configure", "--disable-dependency-tracking", "--prefix=#{prefix}", "--enable-opencl"
+    else
+      system "./configure", "--disable-dependency-tracking", "--prefix=#{prefix}"
+    end
     system "make", "install"
+    if build.with? "training-tools"
+      system "make training"
```
Comment (DomT4) [inline Library/Formula/tesseract.rb]: `system "make", "training"`

File: Library/Formula/tesseract.rb [hunk 5]
```diff
@@ -58,8 +70,16 @@ def install
...
+      system "make training-install"
```
Comment (DomT4) [inline Library/Formula/tesseract.rb]: `system "make", "training-install"`

**Comment by tdsmith** [comment]
OpenCL is supported on 10.6+ so let's turn it on by default.

**Comment by ryanfb** [comment]
@tdsmith While OpenCL is supported in OS X, I would say it's more experimental in Tesseract from my experience, so I'd rather sacrifice it as an option than turn it on by default.

File: Library/Formula/tesseract.rb (hunk 2 above)
Comment (ryanfb) [inline Library/Formula/tesseract.rb]: There appears to be a change in Tesseract HEAD somewhere along the way (since 3.04.00) that introduced a warning about Cairo during the default `./configure` step (I believe it's only used for training-tools though).

File: Library/Formula/tesseract.rb (hunk 1 above)
Comment (ryanfb) [inline Library/Formula/tesseract.rb]: If you want to use Tesseract with anything other than pre-built `.traineddata` etc. files, you need the training tools. I don't have a sense for the general need for that, but a lot of the discussion I see on [the Tesseract-OCR Google Group](https://groups.google.com/forum/#!forum/tesseract-ocr) is about custom OCR that needs training tools.

The other issue may be that due to the way some of the training tools use Pango/Cairo font rendering there seem to be a lot of issues with them on OS X vs. Linux. Part of why I'd like to have this as an option is so that people on OS X who *do* need to do OCR training can start to report those issues upstream in Tesseract and start getting them fixed.

File: Library/Formula/tesseract.rb (hunk 2 above)
Comment (tdsmith) [inline Library/Formula/tesseract.rb]: Cairo is unlikely to be a :build dependency; the shared libraries are probably needed once the tools are installed.

File: Library/Formula/tesseract.rb (hunk 3 above)
Comment (ryanfb) [inline Library/Formula/tesseract.rb]: Done and squashed/pushed.

File: Library/Formula/tesseract.rb (hunk 4 above)
Comment (ryanfb) [inline Library/Formula/tesseract.rb]: Done and squashed/pushed.

File: Library/Formula/tesseract.rb (hunk 5 above)
Comment (ryanfb) [inline Library/Formula/tesseract.rb]: Done and squashed/pushed.

File: Library/Formula/tesseract.rb (hunk 2 above)
Comment (ryanfb) [inline Library/Formula/tesseract.rb]: Removed and squashed/pushed - seems like the warning only occurs during a `./configure` issue with default clang + OpenMP on current Tesseract HEAD (see tesseract-ocr/tesseract#76).

File: Library/Formula/tesseract.rb (hunk 1 above)
Comment (DomT4) [inline Library/Formula/tesseract.rb]: I see. I'm not super-opposed to the option, I guess, it's just the rather heavy dependencies involved.

Sadly, the way `brew` handles `if build.with?` dependencies is less than ideal and consequently we've been trying to get rid of `if build.with?` outside of `def install`. Essentially the option looks like it doesn't require anything extra under `brew info tesseract`:

```
==> Dependencies
Required: leptonica ✘
Recommended: libtiff ✔
==> Options
--all-languages
	Install recognition data for all languages
--with-opencl
	Enable OpenCL support
--with-training-tools
	Install OCR training tools
```

But obviously, there's actually a bunch of "silent" dependencies there when the option is invoked. I'll let the other maintainers weigh in on this particular use case and see what the consensus is.

Comment (ryanfb) [inline Library/Formula/tesseract.rb]: Ah, ok. It wasn't clear to me from [the Homebrew example formula](https://github.com/Homebrew/homebrew/blob/master/Library/Contributions/example-formula.rb) if there was a better way of doing `depends_on` for options.

**Comment by mikemcquaid** [comment]
Let's leave this open for a while to see if people :+1:

**Comment by callmewhy** [comment]
`--with-training-tools` is really necessary if you want to train your own traineddata. I look for this anwser for a long time...Hope it can be mergeed soon : )

**Comment by mikemcquaid** [comment]
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!
